### ISSUE TITLE
Fine tuned Mistral inference issue for >4k context length

### ISSUE THREAD

**Issue by oooodoori** [huggingface/transformers#27966]
**System Info**
- `transformers` version: 4.36.0-dev (main branch)
- Huggingface_hub version: 0.19.4
- PyTorch version: 2.1.0
- Using GPU in script?: yes

We fine tuned a mistralai/Mistral-7B-Instruct-v0.1using LoRa on some 8k context length data. The inferencing was fine with transformers at 4.34.0, but after updating the version, the inferencing became irrelevant repeition for token length > 4096.

We were able to get around this by disabling fast attention 2, but the overall model performance suffered. It seems to be a problem related to the 4D attention mask implementation in transformers 4.35+. This only happens when the token length exceeds 4k. Any ideas what might be wrong?

**Comment by amyeroberts** [comment]
Hi @oooodoori, thanks for raising this issue! 

So that we can best help, could you provide some more details about the issue? e.g.: 
* Have you tried this on 4.35 as well? How about the recent stable 4.36 release?
* Could you provide an example of a good versus bad output? 
* Is the only difference between getting good and bad outputs the transformers version? 
* Could you share a code snippet we can run to reproduce running the inference? Do you see this behaviour on pretrained weights as well e.g. [mistralai/Mistral-7B-v0.1](https://huggingface.co/mistralai/Mistral-7B-v0.1) or only with your own finetuned model? 
* Can you provide more details about how the model was trained e.g. a training script to reproduce including the lora configuration?
